lib/validators/resolveNestedDefinitions.ts

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+import {
+  JsonModel, JsonDefinitions, JsonSpec, JsonParameters, JsonOperation, JsonParameter
+} from "yasway"
+import { updateProperty } from "../util/updateProperty"
+import { objectMap } from "../util/mapObject"
+import { methods } from "../util/methods"
+
+export function resolveNestedDefinitions(spec: JsonSpec): JsonDefinitions {
+  const result = new Result()
+  const definitions = spec.definitions
+  if (definitions) {
+    for (const name in definitions) {
+      const definition = definitions[name]
+      result.extra[name] = definition
+      result.resolveDefinition(name, definition)
+    }
+  }
+  result.resolveParameterMap(spec, "parameters")
+  result.resolveParameterMap(spec, "responses")
+  const paths = spec.paths
+  if (paths) {
+    let i = 0
+    for (const url in paths) {
+      const path = paths[url]
+      const pathName = `paths_${i}`
+      result.resolveParameterArray(pathName, path.parameters)
+      for (const method of methods) {
+        const operation = path[method]
+        if (operation !== undefined) {
+          result.resolveParameterArray(`operations.${operation.operationId}`, operation.parameters)
+        }
+      }
+      ++i
+    }
+  }
+  return result.extra
+}
+
+class Result {
+
+  public readonly extra: JsonDefinitions = {}
+
```

Comment:
not sure why we need a mutable state in the result. can we just have the extra map outside of the result and we just append to it everytime we resolve a defintiion ? also it seems we are having all the definitions there not just the extra ones.

Reply:
something like `let allDefinitions = [...allDefinitions, resolvedDefinitions in `resolveNestedDefinitions` 

---
In reply to: [197909291](https://github.com/Azure/oav/pull/265#discussion_r197909291) [](ancestors = 197909291)